code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/process/LoadBalancerServiceCreate.java

```diff
@@ -0,0 +1,185 @@
+                LOAD_BALANCER.REMOVED, null, LOAD_BALANCER.ACCOUNT_ID, service.getAccountId());
+        if (lb == null) {
+            Map<String, Object> data = new HashMap<>();
+            data.put("name", lbName);
+            data.put(LoadBalancerConstants.FIELD_LB_CONFIG_ID, lbConfig.getId());
+            data.put(LoadBalancerConstants.FIELD_LB_SERVICE_ID, service.getId());
+            data.put(LoadBalancerConstants.FIELD_LB_NETWORK_ID, sdService.getServiceNetworkId(service));
+            data.put(
+                    LoadBalancerConstants.FIELD_LB_INSTANCE_IMAGE_UUID,
+                    launchConfigData.get(InstanceConstants.FIELD_IMAGE_UUID));
+            data.put(
+                    LoadBalancerConstants.FIELD_LB_INSTANCE_URI_PREDICATE,
+                    DataAccessor.fields(service).withKey(LoadBalancerConstants.FIELD_LB_INSTANCE_URI_PREDICATE)
+                            .withDefault("delegate:///").as(
+                                    String.class));
+            data.put("accountId", service.getAccountId());
+            lb = objectManager.create(LoadBalancer.class, data);
+        }
+
+        objectProcessManager.executeProcess(LoadBalancerConstants.PROCESS_LB_CREATE, lb, null);
+    }
+
+    private LoadBalancerConfig createDefaultLoadBalancerConfig(String defaultName,
+            Map<String, Object> lbConfigData, Service service) {
+        String name = lbConfigData.get("name") == null ? defaultName : lbConfigData.get("name")
+                .toString();
+        LoadBalancerConfig lbConfig = objectManager.findOne(LoadBalancerConfig.class,
+                LOAD_BALANCER_CONFIG.REMOVED, null,
+                LOAD_BALANCER_CONFIG.ACCOUNT_ID, service.getAccountId(),
+                LOAD_BALANCER_CONFIG.SERVICE_ID, service.getId());
+
+        if (lbConfig == null) {
+            lbConfigData.put("accountId", service.getAccountId());
+            lbConfigData.put("name", name);
+            lbConfigData.put("serviceId", service.getId());
+            lbConfig = resourceDao.createAndSchedule(LoadBalancerConfig.class, lbConfigData);
+        }
+        return lbConfig;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void createListeners(Service service, LoadBalancerConfig lbConfig, Map<String, Object> launchConfigData) {
+        // 1. create listeners
+        List<String> portDefs = (List<String>) launchConfigData.get(InstanceConstants.FIELD_PORTS);
+        if (portDefs == null || portDefs.isEmpty()) {
+            return;
+        }
+
+        Map<Integer, LoadBalancerListener> listeners = new HashMap<>();
+
+        for (String port : portDefs) {
+            PortSpec spec = new PortSpec(port);
+            if (!port.contains("tcp")) {
+                // default to http unless defined otherwise in the compose file
+                spec.setProtocol("http");
+            }
+
+            if (listeners.containsKey(spec.getPrivatePort())) {
+                continue;
```

Comment:
:) it can happen with idempotent checks in. As with this check on, the loop will never gonna have sequential successes for all the ports defined in the portDefs